### PR TITLE
ffmpeg-qsv: add tcbrc enc tests

### PR DESCRIPTION
**Pull request by uartie**
This is an example for discussion at https://github.com/intel/vaapi-fits/pull/431#discussion_r941994192

cc: @Wanghangjie8

### CODE REVIEW

**Comment by uartie** [comment]
@FocusLuo according to your explanation of TCBRC at https://github.com/intel/vaapi-fits/pull/431#issuecomment-1204644865 this should be a simple way to implement.  Let me know if this makes sense and satisfies the TCBRC requirement.

File: test/ffmpeg-qsv/encode/10bit/hevc.py
Comment (uartie) [inline test/ffmpeg-qsv/encode/10bit/hevc.py:205]: We probably want to create a new, TCBRC variant specific case generator for this instead of using spec from vbr_lp pre-existing cases.

File: test/ffmpeg-qsv/encode/avc.py
Comment (uartie) [inline test/ffmpeg-qsv/encode/avc.py:204]: We probably want to create a new, TCBRC variant specific case generator for this.

File: test/ffmpeg-qsv/encode/hevc.py
Comment (uartie) [inline test/ffmpeg-qsv/encode/hevc.py:205]: We probably want to create a new, TCBRC variant specific case generator for this.

**Comment by FocusLuo** [comment]
> @FocusLuo according to your explanation of TCBRC at [#431 (comment)](https://github.com/intel/vaapi-fits/pull/431#issuecomment-1204644865) this should be a simple way to implement. Let me know if this makes sense and satisfies the TCBRC requirement.

Great, thanks Artie,

**Review by FocusLuo — APPROVED** [review]
LGTM

**Comment by FocusLuo** [comment]
@uartie  not sure your PR will also include a new framesize check (maybe after bitrate check) in assert phase for encoded bitstream

You can use "ffprobe -show_frames -i input.264" to detect output video, which will print "pict_type=xxx" and "pkt_size=xxx" for every frame.
We need check every frame's pkt_size <= bitrate/(8*fps).

**Comment by uartie** [comment]
> @uartie not sure your PR will also include a new framesize check (maybe after bitrate check) in assert phase for encoded bitstream
> 
> You can use "ffprobe -show_frames -i input.264" to detect output video, which will print "pict_type=xxx" and "pkt_size=xxx" for every frame. We need check every frame's pkt_size <= bitrate/(8*fps).

Yes, that is a good idea.

**Comment by FocusLuo** [comment]
@uartie Would you please change this PR status and submit this PR? We need this for ffmpeg-qsv and gst-msdk cloud game testing.

**Comment by uartie** [comment]
> @uartie Would you please change this PR status and submit this PR? We need this for ffmpeg-qsv and gst-msdk cloud game testing.

This still needs a little more work.  It was meant to unblock @Wanghangjie8 to show how to get started.  If you still would like me to finish it, then let me know and give me a couple days.

**Comment by FocusLuo** [comment]
> > @uartie Would you please change this PR status and submit this PR? We need this for ffmpeg-qsv and gst-msdk cloud game testing.
> 
> This still needs a little more work. It was meant to unblock @Wanghangjie8 to show how to get started. If you still would like me to finish it, then let me know and give me a couple days.

Yes, please go ahead and to finish it. It is no problem to use some days. I synced with Hangjie before, he also like yours to be merged.
BTW, gst-msdk also needs this testing. not sure whether you have time and would like to supply a patch after this.

**Comment by uartie** [comment]
> > > @uartie Would you please change this PR status and submit this PR? We need this for ffmpeg-qsv and gst-msdk cloud game testing.
> > 
> > 
> > This still needs a little more work. It was meant to unblock @Wanghangjie8 to show how to get started. If you still would like me to finish it, then let me know and give me a couple days.
> 
> Yes, please go ahead and to finish it. It is no problem to use some days. I synced with Hangjie before, he also like yours to be merged. BTW, gst-msdk also needs this testing. not sure whether you have time and would like to supply a patch after this.

Sure, I can do it.  Please assign the JIRA to me.

File: lib/ffmpeg/qsv/encoder.py
Comment (wenbinc-Bin) [inline lib/ffmpeg/qsv/encoder.py:75]: if max_frame_size_i and max_frame_size_p is not defined, all frames should obey the average frame size. Not only P frames.

Comment (uartie) [reply lib/ffmpeg/qsv/encoder.py:75]: In the cases I tested, I-frames were always too big and sometimes P/B frames are, too.  An "average" of all frame sizes might pass, though.

The check here is not "average"... it strictly checks **each** `frame size <= threshold`.  Should the check be `average(all frame sizes) <= threshold`, instead?

Comment (wenbinc-Bin) [reply lib/ffmpeg/qsv/encoder.py:75]: I double check the VPL's code
1. Sorry I miss one option "-strict -1". To use TCBRC we need to disable NalHrdConformance. This is not listed in doc. I see this limitation in code [https://github.com/oneapi-src/oneVPL-intel-gpu/blob/4c41eeff76ab594575fe3b7519df478ee6d2fd7d/_studio/mfx_lib/encode_hw/hevc/agnostic/base/hevcehw_base_legacy.cpp#L293](https://github.com/oneapi-src/oneVPL-intel-gpu/blob/4c41eeff76ab594575fe3b7519df478ee6d2fd7d/_studio/mfx_lib/encode_hw/hevc/agnostic/base/hevcehw_base_legacy.cpp#L293 )
2. I think we should check `average(all frame) < threshold` and the target average frame size is calculated by "`targetkbps * 1000 / 8 / framerate`" .
3. low_delay_brc also automatically set max_frame_size limitation. max_frame_size is calculated based on maxKbps, but vary on different codec. For h264, it is (`maxkbps * 1000 / 8 / framerate`)[https://github.com/oneapi-src/oneVPL-intel-gpu/blob/4c41eeff76ab594575fe3b7519df478ee6d2fd7d/_studio/mfx_lib/encode_hw/h264/src/mfx_h264_enc_common_hw.cpp#L5204](https://github.com/oneapi-src/oneVPL-intel-gpu/blob/4c41eeff76ab594575fe3b7519df478ee6d2fd7d/_studio/mfx_lib/encode_hw/h264/src/mfx_h264_enc_common_hw.cpp#L5204). For hevc, it is (`1.4 * maxkbps * 1000 / 8 / framerate`) [https://github.com/oneapi-src/oneVPL-intel-gpu/blob/4c41eeff76ab594575fe3b7519df478ee6d2fd7d/_studio/mfx_lib/encode_hw/hevc/agnostic/base/hevcehw_base_max_frame_size.cpp#L102](https://github.com/oneapi-src/oneVPL-intel-gpu/blob/4c41eeff76ab594575fe3b7519df478ee6d2fd7d/_studio/mfx_lib/encode_hw/hevc/agnostic/base/hevcehw_base_max_frame_size.cpp#L102). To see the exact value, you can see ffmpeg log when run command with "-v verbose" option. To test this feature, I suggest to test "`(each frame size) < max_frame_size`" then check "`(failed frame number)/(total frame number) < threshold`". I don't know if we should check this as well, it seems relate to low_delay_brc but doesn't belong to TCBRC.

Comment (uartie) [reply lib/ffmpeg/qsv/encoder.py]: Ok, I updated PR to 1. and 2.  I still get a lot of test cases fail the TCBRC check.  Please try on your side and advise.  You'll need to define some cases on your side as I have not created a patch for them.

Comment (wenbinc-Bin) [reply lib/ffmpeg/qsv/encoder.py]: I tried on my side. many failed case as well. Most of them is caused by low bitrate. I will check with VPL team. Seems I misunderstand some TCBRC's behavior.

Comment (wenbinc-Bin) [reply lib/ffmpeg/qsv/encoder.py]: I confirmed with VPL team.
1. "targetKbps*1000/8/fps" is a reference number. Driver try to make pkt_size as close to this number as possible. When VPL team test this feature they set a boundary like this "`each pkt_size < 1.2 * target_frame_size`". "`1.2`" is not a official number. It depends on video content. I think we can adjust it according to our CI test result.
2. Even we increase the boundary. There are some frame bigger than this number, for example I frame. VPL team said it is expected. We can set another check "`failed frame number/ total frame number < tolerance rate`".
3. According to VPL team, bitrate too low is probably a bug. I will submit a ticket to VPL team.

Comment (uartie) [reply lib/ffmpeg/qsv/encoder.py]: Updated.  Please try now.

Comment (wenbinc-Bin) [reply lib/ffmpeg/qsv/encoder.py]: except the bitrate problem, It works on my side.

Comment (wenbinc-Bin) [reply lib/ffmpeg/qsv/encoder.py]: LGTM, Thanks a lot for you working😄. Sorry for bothering you to change so many times.

**Comment by wenbinc-Bin** [comment]
low bitrate issue: [https://github.com/oneapi-src/oneVPL-intel-gpu/issues/239](https://github.com/oneapi-src/oneVPL-intel-gpu/issues/239)

**Pull request merged**
